Modules/_sqlite/connection.c

```diff
@@ -1839,17 +1839,22 @@ pysqlite_connection_exit_impl(pysqlite_Connection *self, PyObject *exc_type,
                               PyObject *exc_value, PyObject *exc_tb)
 /*[clinic end generated code: output=0705200e9321202a input=bd66f1532c9c54a7]*/
 {
-    const char* method_name;
+    int commit = 0;
     PyObject* result;
 
     if (exc_type == Py_None && exc_value == Py_None && exc_tb == Py_None) {
-        method_name = "commit";
+        commit = 1;
+        result = pysqlite_connection_commit_impl(self);
     } else {
-        method_name = "rollback";
+        result = pysqlite_connection_rollback_impl(self);
     }
 
-    result = PyObject_CallMethod((PyObject*)self, method_name, NULL);
     if (!result) {
+        /* If commit failed, rollback */
+        if (commit) {
+            result = pysqlite_connection_rollback_impl(self);
+            Py_XDECREF(result);
+        }
```

Comment:
Alternative approach where we return `False` if rollback succeeds, iso. just returning `NULL`. I don't think this really matters, the exception is raised either way.

```suggestion
        if (commit) {
            result = pysqlite_connection_rollback_impl(self);
        }
        if (result == NULL) {
            return NULL;
        }
```

Reply:
You cannot return a valid python object with an exception being set. Maybe i am misunderstanding what you mention with "return False"

Reply:
This `__exit__` method defaults to returning `Py_RETURN_FALSE` a few lines below. So `result != NULL` => `Py_RETURN_FALSE`, and `result == NULL` => `return NULL`.

Currently this PR returns `NULL` if commit fails but rollback succeeds. Alternatively we can use `Py_RETURN_FALSE` if commit fails but rollback succeeds. I'm not sure it matters.

Reply:
> Currently this PR returns `NULL` if commit fails but rollback succeeds. Alternatively we can use `Py_RETURN_FALSE` if commit fails but rollback succeeds. I'm not sure it matters.

As long as we are always returning NULL with an exception set, then we are fine.

Reply:
`pysqlite_connection_rollback_impl` can call `PyErr_Clear`, so I added a fix for that. If rollback also fails, we overwrite the current exception with `PyErr_SetString`.